aten/src/ATen/native/vulkan/Vulkan.cpp

```diff
@@ -0,0 +1,1258 @@
+  VkImageViewCreateInfo info{};
+  info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+  info.image = image_;
+  info.viewType = kImageViewType;
+  info.format = kFormat;
+  info.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+  info.subresourceRange.baseMipLevel = 0;
+  info.subresourceRange.levelCount = 1;
+  info.subresourceRange.baseArrayLayer = 0;
+  info.subresourceRange.layerCount = 1;
+  return info;
+}
+
+VkSamplerCreateInfo VImage::makeSamplerCreateInfo() const {
+  VkSamplerCreateInfo info{};
+  info.sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO;
+  info.magFilter = kFilter;
+  info.minFilter = kFilter;
+  info.addressModeU = kSamplerAddressMode;
+  info.addressModeV = kSamplerAddressMode;
+  info.addressModeW = kSamplerAddressMode;
+  info.anisotropyEnable = VK_FALSE;
+  info.maxAnisotropy = 1.0f;
+  info.borderColor = VK_BORDER_COLOR_FLOAT_TRANSPARENT_BLACK;
+  info.compareEnable = VK_FALSE;
+  info.mipmapMode = VK_SAMPLER_MIPMAP_MODE_LINEAR;
+  info.mipLodBias = 0.0f;
+  info.minLod = 0.0f;
+  info.maxLod = 0.0f;
+  return info;
+}
+
+VkDescriptorImageInfo VImage::makeDescriptorImageInfo(
+    VkImageLayout imageLayout) const {
+  VkDescriptorImageInfo info{};
+  info.sampler = sampler_;
+  info.imageView = imageView_;
+  info.imageLayout = imageLayout;
+  return info;
+}
+
+VkWriteDescriptorSet VImage::makeWriteDescriptorSet(
+    VkDescriptorSet descriptorSet,
+    uint32_t binding,
+    VkDescriptorType descriptorType,
+    const VkDescriptorImageInfo* imageInfo) const {
+  VkWriteDescriptorSet writeSet{};
+  writeSet.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+  writeSet.pNext = nullptr;
+  writeSet.dstSet = descriptorSet;
+  writeSet.dstBinding = binding;
+  writeSet.dstArrayElement = 0;
+  writeSet.descriptorCount = 1;
+  writeSet.descriptorType = descriptorType, writeSet.pImageInfo = imageInfo;
+  writeSet.pBufferInfo = nullptr;
+  writeSet.pTexelBufferView = nullptr;
+  return writeSet;
+}
+
+void VImage::bind(
+    VkDescriptorSet descriptorSet,
+    uint32_t binding,
+    VkDescriptorType descriptorType,
+    VkImageLayout imageLayout) const {
+  auto descrImageInfo = makeDescriptorImageInfo(imageLayout);
+  auto writeDescrSet = makeWriteDescriptorSet(
+      descriptorSet, binding, descriptorType, &descrImageInfo);
+  vkUpdateDescriptorSets(context().device(), 1, &writeDescrSet, 0, nullptr);
+}
+
+void VImage::bindShaderRead(VkDescriptorSet descriptorSet, uint32_t binding)
+    const {
+  bind(
+      descriptorSet,
+      binding,
+      VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+      VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
+}
+
+void VImage::bindStorageImage(VkDescriptorSet descriptorSet, uint32_t binding)
+    const {
+  bind(
+      descriptorSet,
+      binding,
+      VK_DESCRIPTOR_TYPE_STORAGE_IMAGE,
+      VK_IMAGE_LAYOUT_GENERAL);
+}
+
+void VImage::addImageMemoryBarrier(
+    VkCommandBuffer commandBuffer,
+    VkImageLayout newLayout) const {
```

Comment:
To be able for const `VulkanTensor` that returns const `VImage` to set memory barrier to use this image for shader read operation (previously it could be used as image store)